client/allocrunner/taskrunner/script_check_hook.go

```diff
@@ -248,38 +239,83 @@ type heartbeater interface {
 type scriptCheck struct {
 	id          string
 	agent       heartbeater
+	check       *structs.ServiceCheck
 	lastCheckOk bool // true if the last check was ok; otherwise false
 	tasklet
 }
 
 // scriptCheckConfig is a parameter struct for newScriptCheck
 type scriptCheckConfig struct {
-	allocID  string
-	taskName string
-	service  *structs.Service
-	check    *structs.ServiceCheck
-	agent    heartbeater
+	allocID    string
+	taskName   string
+	service    *structs.Service
+	check      *structs.ServiceCheck
+	agent      heartbeater
+	driverExec tinterfaces.ScriptExecutor
+	taskEnv    *taskenv.TaskEnv
+	logger     log.Logger
+	shutdownCh chan struct{}
 }
 
 // newScriptCheck constructs a scriptCheck. we're only going to
 // configure the immutable fields of scriptCheck here, with the
 // rest being configured during the Poststart hook so that we have
 // the rest of the task execution environment
 func newScriptCheck(config *scriptCheckConfig) *scriptCheck {
-	serviceID := agentconsul.MakeTaskServiceID(
-		config.allocID, config.taskName, config.service)
-	checkID := agentconsul.MakeCheckID(serviceID, config.check)
+
+	// Guard against not having a valid taskEnv. This can be the case if the
+	// PreKilling or Exited hook is run before Poststart.
+	if config.taskEnv == nil || config.driverExec == nil {
+		return nil
+	}
 
 	sc := &scriptCheck{
-		id:          checkID,
 		agent:       config.agent,
+		check:       config.check,
 		lastCheckOk: true, // start logging on first failure
 	}
+
 	// we can't use the promoted fields of tasklet in the struct literal
-	sc.Command = config.check.Command
-	sc.Args = config.check.Args
+	sc.Command = config.taskEnv.ReplaceEnv(config.check.Command)
+	sc.Args = config.taskEnv.ParseAndReplace(config.check.Args)
 	sc.Interval = config.check.Interval
 	sc.Timeout = config.check.Timeout
+	sc.exec = config.driverExec
+	sc.callback = newScriptCheckCallback(sc)
+	sc.logger = config.logger
+	sc.shutdownCh = config.shutdownCh
+
+	// the hash of the interior structs.ServiceCheck is used by the
+	// Consul client to get the ID to register for the check. So we
+	// update it here so that we have the same ID for UpdateTTL.
+	orig := config.check
+	sc.check.Name = config.taskEnv.ReplaceEnv(orig.Name)
+	sc.check.Type = config.taskEnv.ReplaceEnv(orig.Type)
+	sc.check.Command = sc.Command
+	sc.check.Args = sc.Args
+	sc.check.Path = config.taskEnv.ReplaceEnv(orig.Path)
+	sc.check.Protocol = config.taskEnv.ReplaceEnv(orig.Protocol)
+	sc.check.PortLabel = config.taskEnv.ReplaceEnv(orig.PortLabel)
+	sc.check.InitialStatus = config.taskEnv.ReplaceEnv(orig.InitialStatus)
+	sc.check.Method = config.taskEnv.ReplaceEnv(orig.Method)
+	sc.check.GRPCService = config.taskEnv.ReplaceEnv(orig.GRPCService)
```

Comment:
For the hashes to be the same in the face of interpolation both this taskenv and the taskenv used by the services hook must be the same. I think that's true because the taskenv is sent to the Update hooks. Might be worth a comment of warning or shared helper func to ensure we're doing the exact same interpolation?

Reply:
I think that might be where I'm missing something that's falling down on updates right now in the e2e tests. Will start there next.

Reply:
Partially addressed in da342e2, the rest will be covered in #6272